src/oer/de.rs

```diff
@@ -823,35 +825,40 @@ impl crate::Decoder for Decoder<'_> {
         crate::der::de::Decoder::parse_date_string(&string)
     }
 
-    fn decode_set<FIELDS, SET, D, F>(
+    fn decode_set<const RC: usize, const EC: usize, FIELDS, SET, D, F>(
         &mut self,
         _: Tag,
         decode_fn: D,
         field_fn: F,
     ) -> Result<SET, Self::Error>
     where
-        SET: Decode + Constructed,
+        SET: Decode + Constructed<RC, EC>,
         FIELDS: Decode,
-        D: Fn(&mut Self, usize, Tag) -> Result<FIELDS, Self::Error>,
+        D: Fn(&mut Self::AnyDecoder<RC, EC>, usize, Tag) -> Result<FIELDS, Self::Error>,
         F: FnOnce(Vec<FIELDS>) -> Result<SET, Self::Error>,
     {
-        let (bitmap, extensible_present) = self.parse_preamble::<SET>()?;
+        let (bitmap, extensible_present) = self.parse_preamble::<RC, EC, SET>()?;
 
         let field_map = SET::FIELDS
+            .canonised()
             .optional_and_default_fields()
             .zip(bitmap.into_iter().map(|b| *b))
             .collect::<alloc::collections::BTreeMap<_, _>>();
 
         let decoder_fields = SET::FIELDS
+            .canonised()
             .optional_and_default_fields()
             .zip(bitmap.into_iter().map(|b| *b))
             .collect();
```

Comment:
These two are the same except for the collect at the end right? Wouldn't it make sense to set up the iter once, and then clone it?

Reply:
I improved the function. I have to take a second look for whole decoding implementation later on if I can use the new constants to improve decoding speed further.